File: app/src/main/java/com/tobiasfried/rasa/DetailActivity.java
package com.tobiasfried.rasa;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TextView;

import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.FirebaseFirestore;
import com.tobiasfried.rasa.model.Brew;
import com.tobiasfried.rasa.model.Ingredient;
import com.tobiasfried.rasa.utils.TimeUtility;
import com.tobiasfried.rasa.viewmodel.EntryViewModel;
import com.tobiasfried.rasa.viewmodel.EntryViewModelFactory;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.content.res.ResourcesCompat;
import androidx.lifecycle.ViewModelProviders;
import butterknife.BindView;
import butterknife.ButterKnife;

import static com.tobiasfried.rasa.EntryActivity.EXTRA_BREW_ID;
import static com.tobiasfried.rasa.EntryActivity.EXTRA_BREW_ID_HISTORY;

public class DetailActivity extends AppCompatActivity {

    // Database
    protected FirebaseFirestore db;
    protected EntryViewModel viewModel;
    protected String brewId;
    protected String collection = Brew.CURRENT;
    protected DocumentReference docRef;

    // Model
    protected Brew currentBrew;
    protected List<Ingredient> teas;
    private List<Ingredient> ingredients;
    private List<Ingredient> selectedIngredients;

    // Views
    @BindView(R.id.create_edit_text)
    EditText brewNameEditText;

    @BindView(R.id.primary_date_calendar)
    TextView primaryDateTextView;

    @BindView(R.id.primary_remaining_days)
    TextView primaryRemainingDaysTextView;

    @BindView(R.id.tea)
    LinearLayout teaList;

    @BindView(R.id.tea_name_autocomplete)
    AutoCompleteTextView teaNameEditText;

    @BindView(R.id.tea_amount_picker)
    EditText teaAmountEditText;

    @BindView(R.id.tea_picker)
    Spinner teaSpinner;

    @BindView(R.id.primary_sugar_picker)
    Spinner primarySugarSpinner;

    @BindView(R.id.sugar_amount_picker)
    EditText primarySugarAmountEditText;

    @BindView(R.id.water_amount_picker)
    EditText waterAmountEditText;

    @BindView(R.id.secondary_date_calendar)
    TextView secondaryDateTextView;

    @BindView(R.id.secondary_remaining_days)
    TextView secondaryRemainingDaysTextView;

    @BindView(R.id.secondary_sugar_picker)
    Spinner secondarySugarSpinner;

    @BindView(R.id.secondary_sugar_amount_picker)
    EditText secondarySugarAmountEditText;

    @BindView(R.id.ingredient_chip_group)
    ChipGroup flavorChipGroup;

    @BindView(R.id.notes)
    EditText notesEditText;

    private boolean editing = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail);
        ButterKnife.bind(this);

        // TODO fetch from loaded brew
        selectedIngredients = new ArrayList<>();

        // Get Database instance
        db = FirebaseFirestore.getInstance();

        // Get Toolbar
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.inflateMenu(R.menu.menu_entry);

        // Get ViewModel
        if (getIntent().hasExtra(EXTRA_BREW_ID)) {
            brewId = Objects.requireNonNull(getIntent().getExtras()).getString(EXTRA_BREW_ID);
        } else if (getIntent().hasExtra(EXTRA_BREW_ID_HISTORY)) {
            brewId = Objects.requireNonNull(getIntent().getExtras()).getString(EXTRA_BREW_ID_HISTORY);
            collection = Brew.HISTORY;
            // TODO Not this
        }
        EntryViewModelFactory factory = new EntryViewModelFactory(db, collection, brewId);
        viewModel = ViewModelProviders.of(this, factory).get(EntryViewModel.class);
        fetchBrew();

        docRef = viewModel.getDocumentReference();

        setFieldsEditable();
    }

    private void setFieldsEditable() {
        brewNameEditText.setEnabled(editing);
        teaNameEditText.setEnabled(editing);
        teaSpinner.setEnabled(editing);
        teaAmountEditText.setEnabled(editing);
        primarySugarSpinner.setEnabled(editing);
        primarySugarAmountEditText.setEnabled(editing);
        waterAmountEditText.setEnabled(editing);

        for (int i = 1; i < teaList.getChildCount(); i++) {
            View v = teaList.getChildAt(i);
            v.findViewById(R.id.tea_name_autocomplete).setEnabled(editing);
            v.findViewById(R.id.tea_picker).setEnabled(editing);
            v.findViewById(R.id.tea_amount_picker).setEnabled(editing);
            v.findViewById(R.id.remove_button).setVisibility(editing ? View.VISIBLE : View.GONE);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_entry, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_cancel:
                finish();
        }
        return true;
    }

    private void setupTeaSpinner(Spinner spinner, int position) {
        ArrayAdapter<CharSequence> teaTypes = ArrayAdapter.createFromResource(this, R.array.array_tea_types, R.layout.spinner_item_small);
        teaTypes.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(teaTypes);
        spinner.setSelection(position);
    }

    private void setupSugarSpinners() {
        ArrayAdapter<CharSequence> sweetenerTypes = ArrayAdapter.createFromResource(this, R.array.array_sugar_types, R.layout.spinner_item);
        sweetenerTypes.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        primarySugarSpinner.setAdapter(sweetenerTypes);
        secondarySugarSpinner.setAdapter(sweetenerTypes);
    }

    private void fetchBrew() {
        viewModel.getBrew().observe(this, brew -> {
            currentBrew = brew;
            if (brewId != null) {
                setupBrew();
            } else {
                setupTeaSpinner(teaSpinner, 0);
                setupSugarSpinners();
                refreshDates();
            }
        });
    }

    private void fetchIngredients() {
        viewModel.getFlavors().observe(this, allIngredients -> {
            ingredients = allIngredients;
            setupChips();
        });

        viewModel.getTeas().observe(this, allTeas -> teas = allTeas);
    }

    private void setupBrew() {
        refreshDates();
        fetchIngredients();

        brewNameEditText.setText(currentBrew.getRecipe().getName());

        // Populate Teas
        LayoutInflater inflater = getLayoutInflater();

        Ingredient firstTea = currentBrew.getRecipe().getTeas().get(0);
        teaNameEditText.setText(firstTea.getName());
        setupTeaSpinner(teaSpinner, firstTea.getTeaType().getCode());
        teaAmountEditText.setText(String.valueOf(firstTea.getAmount()));

        for (int i = 1; i < currentBrew.getRecipe().getTeas().size(); i++) {
            Ingredient tea = currentBrew.getRecipe().getTeas().get(i);

            final View teaItem = inflater.inflate(R.layout.list_item_tea, null);

            AutoCompleteTextView nameEditText = teaItem.findViewById(R.id.tea_name_autocomplete);
            nameEditText.setText(tea.getName(), TextView.BufferType.EDITABLE);

            EditText amountEditText = teaItem.findViewById(R.id.tea_amount_picker);
            amountEditText.setText(String.valueOf(tea.getAmount()), TextView.BufferType.EDITABLE);

            Spinner typeSpinner = teaItem.findViewById(R.id.tea_picker);
            setupTeaSpinner(typeSpinner, tea.getTeaType().getCode());

            ImageButton removeButton = teaItem.findViewById(R.id.remove_button);
            removeButton.setVisibility(View.GONE);
            removeButton.setOnClickListener(v -> teaList.removeView(teaItem));

            teaList.addView(teaItem);
        }

        setupSugarSpinners();

        primarySugarSpinner.setSelection(currentBrew.getRecipe().getPrimarySweetener());
        primarySugarAmountEditText.setText(String.valueOf(currentBrew.getRecipe().getPrimarySweetenerAmount()), TextView.BufferType.EDITABLE);

        waterAmountEditText.setText(String.valueOf(currentBrew.getRecipe().getWater()));

        secondarySugarSpinner.setSelection(currentBrew.getRecipe().getSecondarySweetener());
        secondarySugarAmountEditText.setText(String.valueOf(currentBrew.getRecipe().getSecondarySweetenerAmount()), TextView.BufferType.EDITABLE);

        selectedIngredients.addAll(currentBrew.getRecipe().getIngredients());

        notesEditText.setText(currentBrew.getRecipe().getNotes());
    }

    private void setupChips() {
        flavorChipGroup.removeAllViews();
        // Populate ChipGroup from ingredients list
        for (final Ingredient i : currentBrew.getRecipe().getIngredients()) {
            Chip chip = new Chip(this, null, R.style.ChipStyle);
            chip.setText(i.getName().toLowerCase());
            chip.setTextColor(getResources().getColor(R.color.colorAccent, getTheme()));
            chip.setTypeface(ResourcesCompat.getFont(this, R.font.google_sans_medium));
//            chip.setChipBackgroundColorResource(R.color.color_states_chips);
            chip.setChipBackgroundColorResource(android.R.color.transparent);
            chip.setChipStrokeColorResource(R.color.colorAccent);
            chip.setChipStrokeWidth(4.0f);
            chip.setChecked(true);
            chip.setEnabled(false);
            chip.setCheckedIconVisible(false);

            // Add to ChipGroup
            flavorChipGroup.addView(chip);
        }
    }

    private void refreshDates() {
        if (currentBrew.getPrimaryFerment() != null) {
            String primaryDatesString = (currentBrew.getPrimaryFerment().first != null ? TimeUtility.formatDateShort(currentBrew.getPrimaryFerment().first) : "") +
                    " – " + (currentBrew.getPrimaryFerment().second != null ? TimeUtility.formatDateShort(currentBrew.getPrimaryFerment().second) : "");
            primaryDateTextView.setText(primaryDatesString);
            int primaryDays = TimeUtility.daysBetween(currentBrew.getPrimaryFerment().first, currentBrew.getPrimaryFerment().second);
            String primaryDaysString = getResources().getQuantityString(R.plurals.pluralDays, primaryDays, primaryDays);
            primaryRemainingDaysTextView.setText(primaryDaysString);
        }

        if (currentBrew.getSecondaryFerment() != null) {
            String secondaryDatesString = (currentBrew.getSecondaryFerment().first != null ? TimeUtility.formatDateShort(currentBrew.getSecondaryFerment().first) : "") +
                    " – " + (currentBrew.getSecondaryFerment().second != null ? TimeUtility.formatDateShort(currentBrew.getSecondaryFerment().second) : "");
            secondaryDateTextView.setText(secondaryDatesString);
            if (currentBrew.getSecondaryFerment().second != null) {
                int secondaryDays = TimeUtility.daysBetween(currentBrew.getSecondaryFerment().first, currentBrew.getSecondaryFerment().second);
                String secondaryDaysString = getResources().getQuantityString(R.plurals.pluralDays, secondaryDays, secondaryDays);
                secondaryRemainingDaysTextView.setText(secondaryDaysString);
            }
        }
    }

}

File: app/src/main/java/com/tobiasfried/rasa/DetailActivity.kt
package com.tobiasfried.rasa

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.EditText
import android.widget.ImageButton
import android.widget.LinearLayout
import android.widget.Spinner
import android.widget.TextView

import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore
import com.tobiasfried.rasa.domain.Brew
import com.tobiasfried.rasa.domain.Ingredient
import com.tobiasfried.rasa.utils.TimeUtility
import com.tobiasfried.rasa.entry.EntryViewModel
import com.tobiasfried.rasa.entry.EntryViewModelFactory

import java.util.ArrayList
import java.util.Objects

import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.res.ResourcesCompat
import androidx.lifecycle.ViewModelProviders
import butterknife.BindView
import butterknife.ButterKnife

import com.tobiasfried.rasa.entry.EntryActivity.EXTRA_BREW_ID
import com.tobiasfried.rasa.entry.EntryActivity.EXTRA_BREW_ID_HISTORY

class DetailActivity : AppCompatActivity() {

    // Database
    protected var db: FirebaseFirestore
    protected var viewModel: EntryViewModel
    protected var brewId: String? = null
    protected var collection = Brew.CURRENT
    protected var docRef: DocumentReference? = null

    // Model
    protected var currentBrew: Brew
    protected var teas: List<Ingredient>
    private var ingredients: List<Ingredient>? = null
    private var selectedIngredients: MutableList<Ingredient>? = null

    // Views
    @BindView(R.id.create_edit_text)
    internal var brewNameEditText: EditText? = null

    @BindView(R.id.primary_date_calendar)
    internal var primaryDateTextView: TextView? = null

    @BindView(R.id.primary_remaining_days)
    internal var primaryRemainingDaysTextView: TextView? = null

    @BindView(R.id.tea)
    internal var teaList: LinearLayout? = null

    @BindView(R.id.tea_name_autocomplete)
    internal var teaNameEditText: AutoCompleteTextView? = null

    @BindView(R.id.tea_amount_picker)
    internal var teaAmountEditText: EditText? = null

    @BindView(R.id.tea_picker)
    internal var teaSpinner: Spinner? = null

    @BindView(R.id.primary_sugar_picker)
    internal var primarySugarSpinner: Spinner? = null

    @BindView(R.id.sugar_amount_picker)
    internal var primarySugarAmountEditText: EditText? = null

    @BindView(R.id.water_amount_picker)
    internal var waterAmountEditText: EditText? = null

    @BindView(R.id.secondary_date_calendar)
    internal var secondaryDateTextView: TextView? = null

    @BindView(R.id.secondary_remaining_days)
    internal var secondaryRemainingDaysTextView: TextView? = null

    @BindView(R.id.secondary_sugar_picker)
    internal var secondarySugarSpinner: Spinner? = null

    @BindView(R.id.secondary_sugar_amount_picker)
    internal var secondarySugarAmountEditText: EditText? = null

    @BindView(R.id.ingredient_chip_group)
    internal var flavorChipGroup: ChipGroup? = null

    @BindView(R.id.notes)
    internal var notesEditText: EditText? = null

    private val editing = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail)
        ButterKnife.bind(this)

        // TODO fetch from loaded brew
        selectedIngredients = ArrayList()

        // Get Database instance
        db = FirebaseFirestore.getInstance()

        // Get Toolbar
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        toolbar.inflateMenu(R.menu.menu_entry)

        // Get ViewModel
        if (intent.hasExtra(EXTRA_BREW_ID)) {
            brewId = Objects.requireNonNull(intent.extras).getString(EXTRA_BREW_ID)
        } else if (intent.hasExtra(EXTRA_BREW_ID_HISTORY)) {
            brewId = Objects.requireNonNull(intent.extras).getString(EXTRA_BREW_ID_HISTORY)
            collection = Brew.HISTORY
            // TODO Not this
        }
        val factory = EntryViewModelFactory(db, collection, brewId)
        viewModel = ViewModelProviders.of(this, factory).get(EntryViewModel::class.java)
        fetchBrew()

        docRef = viewModel.documentReference

        setFieldsEditable()
    }

    private fun setFieldsEditable() {
        brewNameEditText!!.isEnabled = editing
        teaNameEditText!!.isEnabled = editing
        teaSpinner!!.isEnabled = editing
        teaAmountEditText!!.isEnabled = editing
        primarySugarSpinner!!.isEnabled = editing
        primarySugarAmountEditText!!.isEnabled = editing
        waterAmountEditText!!.isEnabled = editing

        for (i in 1 until teaList!!.childCount) {
            val v = teaList!!.getChildAt(i)
            v.findViewById<View>(R.id.tea_name_autocomplete).isEnabled = editing
            v.findViewById<View>(R.id.tea_picker).isEnabled = editing
            v.findViewById<View>(R.id.tea_amount_picker).isEnabled = editing
            v.findViewById<View>(R.id.remove_button).visibility = if (editing) View.VISIBLE else View.GONE
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_entry, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_cancel -> finish()
        }
        return true
    }

    private fun setupTeaSpinner(spinner: Spinner, position: Int) {
        val teaTypes = ArrayAdapter.createFromResource(this, R.array.array_tea_types, R.layout.spinner_item_small)
        teaTypes.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = teaTypes
        spinner.setSelection(position)
    }

    private fun setupSugarSpinners() {
        val sweetenerTypes = ArrayAdapter.createFromResource(this, R.array.array_sugar_types, R.layout.spinner_item)
        sweetenerTypes.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        primarySugarSpinner!!.adapter = sweetenerTypes
        secondarySugarSpinner!!.adapter = sweetenerTypes
    }

    private fun fetchBrew() {
        viewModel.brew.observe(this, { brew ->
            currentBrew = brew
            if (brewId != null) {
                setupBrew()
            } else {
                setupTeaSpinner(teaSpinner!!, 0)
                setupSugarSpinners()
                refreshDates()
            }
        })
    }

    private fun fetchIngredients() {
        viewModel.flavors.observe(this, { allIngredients ->
            ingredients = allIngredients
            setupChips()
        })

        viewModel.teas.observe(this, { allTeas -> teas = allTeas })
    }

    private fun setupBrew() {
        refreshDates()
        fetchIngredients()

        brewNameEditText!!.setText(currentBrew.recipe.name)

        // Populate Teas
        val inflater = layoutInflater

        val firstTea = currentBrew.recipe.teas[0]
        teaNameEditText!!.setText(firstTea.name)
        setupTeaSpinner(teaSpinner!!, firstTea.teaType!!.code)
        teaAmountEditText!!.setText(firstTea.amount.toString())

        for (i in 1 until currentBrew.recipe.teas.size) {
            val tea = currentBrew.recipe.teas[i]

            val teaItem = inflater.inflate(R.layout.list_item_tea, null)

            val nameEditText = teaItem.findViewById<AutoCompleteTextView>(R.id.tea_name_autocomplete)
            nameEditText.setText(tea.name, TextView.BufferType.EDITABLE)

            val amountEditText = teaItem.findViewById<EditText>(R.id.tea_amount_picker)
            amountEditText.setText(tea.amount.toString(), TextView.BufferType.EDITABLE)

            val typeSpinner = teaItem.findViewById<Spinner>(R.id.tea_picker)
            setupTeaSpinner(typeSpinner, tea.teaType!!.code)

            val removeButton = teaItem.findViewById<ImageButton>(R.id.remove_button)
            removeButton.visibility = View.GONE
            removeButton.setOnClickListener { v -> teaList!!.removeView(teaItem) }

            teaList!!.addView(teaItem)
        }

        setupSugarSpinners()

        primarySugarSpinner!!.setSelection(currentBrew.recipe.primarySweetener)
        primarySugarAmountEditText!!.setText(currentBrew.recipe.primarySweetenerAmount.toString(), TextView.BufferType.EDITABLE)

        waterAmountEditText!!.setText(currentBrew.recipe.water.toString())

        secondarySugarSpinner!!.setSelection(currentBrew.recipe.secondarySweetener)
        secondarySugarAmountEditText!!.setText(currentBrew.recipe.secondarySweetenerAmount.toString(), TextView.BufferType.EDITABLE)

        selectedIngredients!!.addAll(currentBrew.recipe.ingredients)

        notesEditText!!.setText(currentBrew.recipe.notes)
    }

    private fun setupChips() {
        flavorChipGroup!!.removeAllViews()
        // Populate ChipGroup from ingredients list
        for (i in currentBrew.recipe.ingredients) {
            val chip = Chip(this, null, R.style.ChipStyle)
            chip.text = i.name!!.toLowerCase()
            chip.setTextColor(resources.getColor(R.color.colorAccent, theme))
            chip.typeface = ResourcesCompat.getFont(this, R.font.google_sans_medium)
            //            chip.setChipBackgroundColorResource(R.color.color_states_chips);
            chip.setChipBackgroundColorResource(android.R.color.transparent)
            chip.setChipStrokeColorResource(R.color.colorAccent)
            chip.chipStrokeWidth = 4.0f
            chip.isChecked = true
            chip.isEnabled = false
            chip.isCheckedIconVisible = false

            // Add to ChipGroup
            flavorChipGroup!!.addView(chip)
        }
    }

    private fun refreshDates() {
        if (currentBrew.primaryFerment != null) {
            val primaryDatesString = (if (currentBrew.primaryFerment!!.first != null) TimeUtility.formatDateShort(currentBrew.primaryFerment!!.first!!) else "") +
                    " – " + if (currentBrew.primaryFerment!!.second != null) TimeUtility.formatDateShort(currentBrew.primaryFerment!!.second!!) else ""
            primaryDateTextView!!.text = primaryDatesString
            val primaryDays = TimeUtility.daysBetween(currentBrew.primaryFerment!!.first!!, currentBrew.primaryFerment!!.second!!)
            val primaryDaysString = resources.getQuantityString(R.plurals.pluralDays, primaryDays, primaryDays)
            primaryRemainingDaysTextView!!.text = primaryDaysString
        }

        if (currentBrew.secondaryFerment != null) {
            val secondaryDatesString = (if (currentBrew.secondaryFerment!!.first != null) TimeUtility.formatDateShort(currentBrew.secondaryFerment!!.first!!) else "") +
                    " – " + if (currentBrew.secondaryFerment!!.second != null) TimeUtility.formatDateShort(currentBrew.secondaryFerment!!.second!!) else ""
            secondaryDateTextView!!.text = secondaryDatesString
            if (currentBrew.secondaryFerment!!.second != null) {
                val secondaryDays = TimeUtility.daysBetween(currentBrew.secondaryFerment!!.first!!, currentBrew.secondaryFerment!!.second!!)
                val secondaryDaysString = resources.getQuantityString(R.plurals.pluralDays, secondaryDays, secondaryDays)
                secondaryRemainingDaysTextView!!.text = secondaryDaysString
            }
        }
    }

}
